Display user nickname instead of name submitted with form or their email address

anna-giang opened the pull request:
Display the user nickname they set when logging in for the first time as the author name of the associated comment. 
Remove the "Name" form field, as this is now redundant.
Provide the user with the option of showing their email with their comment (a checkbox field on the comment form).

anna-giang committed:
Remove "Name" field from frontend and display logged in user nickname instead

Name field is redundant since only logged in users can access comment form, and all logged in users will have set a nickname.
Instead of fetching commentAuthor from the Name field, fetch the user nickname from the asociated UserInfo entity, and store the user nickname in the commentAuthor field, and display this where the input from the Name field would normally be displayed.

anna-giang committed:
Hide email unless requested by commenter

Add checkbox to comment form for user to choose whether to reveal their email in their comment.
Add backend support in DataServer to store this input field in datastore, and return the input in this field in the response JSON.
Also modify the comments displaying to support this (only display email if user has specified so).

anna-giang committed:
Fix style: add more space between each input field in comment form

anna-giang commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -196,4 +203,22 @@ private String getUploadedFileUrl(HttpServletRequest request, String formInputEl
+   * set a nickname. 
+   * @param id user id of the user.
+   */
+  private String getUserNickname(String id) {
```
The same function is used by both the DataServlet and NicknameServlet, so should I move it into an external class so that I don't have to repeat the implementation in both servlets? 
That seems to be the better practice to avoid repeated code, but not sure what I would name the external class (since it would really just have one method). Or should I make it a public method in the NicknameServlet?

austinct replied:
I think either creating a class called like `Nickname` that has public static functions to deal with getting/setting nicknames from the datastore, or changing `getUserNickname()` to be public static in `NicknameServlet` are fine. Maybe try both and see what you like better? Note you will want these functions to be `static` so you don't have to create an instance of the class to call them. In general helper functions are `static` because you don't need to access any member variables.

anna-giang replied:
I've made a new Nickname class as you suggested, and put the `getUserNickname` function as a static method in there. I think the function is not really related to the the servlet's purpose, so I think a new class is the better way to go. Thankyou!

anna-giang committed:
Move getUserNickname function to new Nickname class

getUserNickname is now a static function in Nickname, a class containing helper functions to facilitate user nicknames.
This function was previously duplicated in both DataServlet and NicknameServlet, which is repeated code and not ideal practice.
The function was not made into a static function in the servlet classes, for it was not considered to be closely related to either servlets' key purpose and operations.

anna-giang commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -70,10 +70,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       String commentText = (String) entity.getProperty("commentText");
       String commentAuthor = (String) entity.getProperty("commentAuthor");
       String imageUrl = (String) entity.getProperty("attachedImage");
+      String showEmail = (String) entity.getProperty("showEmail");
```
The value of showEmail will be passed back to the client in a JSON string, so eventually it will have to be converted back into a string. In this case, should it still be made into a boolean first?

austinct replied:
Even if JSON is sent as a string, JSON can still represent boolean types, so it should be a boolean. Ideally, you'd store it in the datastore as a boolean, and also store it in `comment` as a boolean (the easiest way to do that would be to change `comment` into a `HashMap<String, Object>`). This way, you can always treat `showEmail` as a boolean, even in the javascript code.

anna-giang replied:
Yep okay, that makes sense! Thanks for the feedback and tips 😊

anna-giang committed:
Only send user email in response if user selected to reveal email

For better privacy.
Previously email was included in reponse regardless of the selection, and was just not displayed. However the email could be viewed in Dev Tools, which undos the purpose of the option to show/hide email, which was to provide additional privacy.

anna-giang commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -70,10 +70,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       comment.put("commentAuthor", commentAuthor);
       comment.put("attachedImage", imageUrl); 
+      comment.put("showEmail", showEmail);
       comment.put("authorEmail", email);
```
Yep, that makes sense. Fixed!

anna-giang committed:
Change showEmail field to be boolean instead of string type

showEmail field is stored in Datastore as a boolean and returned by DataServlet as a boolean in the JSON response as well.
This makes more sense as showEmail is inherently a boolean flag.

austinct reviewed (approved):
:+1:
